services/navigation/builtin/builtin.go

```diff
@@ -348,16 +389,28 @@ func (svc *builtIn) nextWaypoint(ctx context.Context) (navigation.Waypoint, erro
 }
 
 func (svc *builtIn) waypointReached(ctx context.Context) error {
-	wp, err := svc.nextWaypoint(ctx)
-	if err != nil {
-		return fmt.Errorf("can't mark waypoint reached: %w", err)
+	if ctx.Err() != nil {
+		return ctx.Err()
+	}
+
+	svc.mu.Lock()
+	defer svc.mu.Unlock()
```

Comment:
There might be latency in the `svc.store` call.
I would recommend caching `svc.waypointInProgress` in a local variable, unlocking the mutex, then calling `WaypointVisited`.

The `store` has its own mutex so you don't need to be protecting that here.